src/ucs/async/thread.c

```diff
@@ -53,26 +54,48 @@ static void ucs_async_thread_hold(ucs_async_thread_t *thread)
 static void ucs_async_thread_put(ucs_async_thread_t *thread)
 {
     if (ucs_atomic_fadd32(&thread->refcnt, -1) == 1) {
-        close(thread->epfd);
+        ucs_event_set_cleanup(thread->event_set);
         ucs_async_pipe_destroy(&thread->wakeup);
         ucs_timerq_cleanup(&thread->timerq);
         ucs_free(thread);
     }
 }
 
+static void ucs_async_thread_ev_handler(int fd, int event,
+                                        ucs_event_set_extra_args **args)
+{
+    ucs_async_thread_t *thread = args[0];
+    int *is_missed = (int*)args[1];
+    ucs_status_t status;
+    ucs_trace_async("ucs_async_thread_ev_handler(fd=%d, event=%d)",
+                    fd, event);
+    if (fd == ucs_async_pipe_rfd(&thread->wakeup)) {
+        ucs_trace_async("progress thread woken up");
+        ucs_async_pipe_drain(&thread->wakeup);
+        return;
+    }
+
+    status = ucs_async_dispatch_handlers(&fd, 1);
+    if (status == UCS_ERR_NO_PROGRESS) {
+         *is_missed = 1;
+    }
+}
+
 static void *ucs_async_thread_func(void *arg)
 {
     ucs_async_thread_t *thread = arg;
-    struct epoll_event events[UCS_ASYNC_EPOLL_MAX_EVENTS];
     ucs_time_t last_time, curr_time, timer_interval, time_spent;
-    int i, nready, is_missed, timeout_ms;
+    int is_missed, timeout_ms;
     ucs_status_t status;
-    int fd;
+    ucs_event_set_extra_args *args[2];
 
     is_missed  = 0;
     curr_time  = ucs_get_time();
     last_time  = ucs_get_time();
 
+    args[0] = thread;
+    args[1] = &is_missed;
```

Comment:
Should I use `ucs_event_set_extra_args *args[] = {thread, &is_missed;}`.
I found `is_missed  = 0;` in the line 92. so It is better to use this syntax.

Reply:
sure, `ucs_event_set_extra_args *args[] = {thread, &is_missed;}` was an example.